akka-cluster-metrics/src/main/scala/akka/cluster/metrics/protobuf/MessageSerializer.scala

```diff
@@ -0,0 +1,219 @@
+    val bos = new ByteArrayOutputStream(BufferSize)
+    val zip = new GZIPOutputStream(bos)
+    msg.writeTo(zip)
+    zip.close()
+    bos.toByteArray
+  }
+
+  def decompress(bytes: Array[Byte]): Array[Byte] = {
+    val in = new GZIPInputStream(new ByteArrayInputStream(bytes))
+    val out = new ByteArrayOutputStream()
+    val buffer = new Array[Byte](BufferSize)
+
+    @tailrec def readChunk(): Unit = in.read(buffer) match {
+      case -1 ⇒ ()
+      case n ⇒
+        out.write(buffer, 0, n)
+        readChunk()
+    }
+
+    readChunk()
+    out.toByteArray
+  }
+
+  def fromBinary(bytes: Array[Byte], clazz: Option[Class[_]]): AnyRef = clazz match {
+    case Some(c) ⇒ fromBinaryMap.get(c.asInstanceOf[Class[ClusterMetricsMessage]]) match {
+      case Some(f) ⇒ f(bytes)
+      case None    ⇒ throw new IllegalArgumentException(s"Unimplemented deserialization of message class $c in metrics")
+    }
+    case _ ⇒ throw new IllegalArgumentException("Need a metrics message class to be able to deserialize bytes in metrics")
+  }
+
+  private def addressFromBinary(bytes: Array[Byte]): Address =
+    addressFromProto(cm.Address.parseFrom(bytes))
+
+  private def addressToProto(address: Address): cm.Address.Builder = address match {
+    case Address(protocol, system, Some(host), Some(port)) ⇒
+      cm.Address.newBuilder().setSystem(system).setHostname(host).setPort(port).setProtocol(protocol)
+    case _ ⇒ throw new IllegalArgumentException(s"Address [${address}] could not be serialized: host or port missing.")
+  }
+
+  private def addressToProtoByteArray(address: Address): Array[Byte] = addressToProto(address).build.toByteArray
+
+  @volatile
+  private var protocolCache: String = null
+  @volatile
+  private var systemCache: String = null
+
+  private def getProtocol(address: cm.Address): String = {
+    val p = address.getProtocol
+    val pc = protocolCache
+    if (pc == p) pc
+    else {
+      protocolCache = p
+      p
+    }
+  }
+
+  private def getSystem(address: cm.Address): String = {
+    val s = address.getSystem
+    val sc = systemCache
+    if (sc == s) sc
+    else {
+      systemCache = s
+      s
+    }
+  }
+
+  private def addressFromProto(address: cm.Address): Address =
+    Address(getProtocol(address), getSystem(address), address.getHostname, address.getPort)
+
+  private def mapWithErrorMessage[T](map: Map[T, Int], value: T, unknown: String): Int = map.get(value) match {
+    case Some(x) ⇒ x
+    case _       ⇒ throw new IllegalArgumentException(s"Unknown ${unknown} [${value}] in cluster message")
+  }
+
+  private def metricsGossipEnvelopeToProto(envelope: MetricsGossipEnvelope): cm.MetricsGossipEnvelope = {
+    val mgossip = envelope.gossip
+    val allAddresses = mgossip.nodes.foldLeft(Set.empty[Address])((s, n) ⇒ s + n.address)
+    val addressMapping = allAddresses.zipWithIndex.toMap
+    val allMetricNames = mgossip.nodes.foldLeft(Set.empty[String])((s, n) ⇒ s ++ n.metrics.iterator.map(_.name))
```

Comment:
@patriknw for the master,merge this first and then I submit a another one,or @Andrei-Pozolotin fix it here.
the #16669 is for release 2.3,see issue here.
https://github.com/akka/akka/issues/16667